File: mobile/ios/Features/QervonFeatures/Sources/CustomerOrderFeature/OrderDetailViewModel.swift
// =============================================================================
// File:           mobile/ios/Features/QervonFeatures/Sources/CustomerOrderFeature/OrderDetailViewModel.swift
// Project:        Qervon
// Author:         USDTG GROUP TECHNOLOGY LLC
// Developer:      Irfan Gedik
// Created Date:   2026-08-12
// Version:        0.1.0
//
// Description:
//   Drives the order detail/tracking screen: polls the courier's last
//   reported location and the ETA endpoint (~5s) while the order is
//   active, and exposes cancel/rate actions. Polling, not a WebSocket
//   subscription — consistent with the Courier app's job-offer polling and
//   much simpler for a single-order view.
//
// License:
//   Qervon License v1.0 — see LICENSE in the repository root.
// =============================================================================

import Foundation
import QervonCore
import QervonNetworking

@MainActor
public final class OrderDetailViewModel: ObservableObject {
    @Published public private(set) var order: Order
    @Published public private(set) var courierLocation: LocationSnapshot?
    @Published public private(set) var eta: EtaInfo?
    @Published public private(set) var isCancelling = false
    @Published public var errorMessage: String?

    private let api: QervonAPI
    private var pollingTask: Task<Void, Never>?
    private let pollingInterval: Duration = .seconds(5)

    public init(order: Order, api: QervonAPI) {
        self.order = order
        self.api = api
    }

    public var isTrackable: Bool {
        order.status == .courierAssigned || order.status == .inTransit
    }

    public var canCancel: Bool {
        order.status == .pending || order.status == .courierAssigned
    }

    public func onAppear() {
        startPolling()
    }

    public func onDisappear() {
        pollingTask?.cancel()
    }

    private func startPolling() {
        pollingTask?.cancel()
        pollingTask = Task { [weak self] in
            guard let self else { return }
            while !Task.isCancelled {
                await self.refresh()
                if !self.isTrackable {
                    return
                }
                try? await Task.sleep(for: self.pollingInterval)
            }
        }
    }

    private func refresh() async {
        guard isTrackable else { return }
        async let locationResult = try? api.getOrderTracking(orderId: order.id)
        async let etaResult = try? api.getOrderEta(orderId: order.id)
        courierLocation = await locationResult ?? nil
        eta = await etaResult ?? nil
    }

    public func cancel() async -> Bool {
        isCancelling = true
        errorMessage = nil
        defer { isCancelling = false }
        do {
            order = try await api.cancelOrder(orderId: order.id)
            pollingTask?.cancel()
            return true
        } catch {
            errorMessage = error.localizedDescription
            return false
        }
    }

    public func rate(stars: Int, comment: String?) async -> Bool {
        errorMessage = nil
        do {
            _ = try await api.rateOrder(orderId: order.id, stars: stars, comment: comment)
            return true
        } catch {
            errorMessage = error.localizedDescription
            return false
        }
    }

    public func openSupportTicket(subject: String, message: String) async -> Bool {
        errorMessage = nil
        do {
            _ = try await api.createSupportTicket(orderId: order.id, subject: subject, message: message)
            return true
        } catch {
            errorMessage = error.localizedDescription
            return false
        }
    }
}

File: mobile/android/feature/customerorder/src/main/kotlin/com/qervon/features/customerorder/OrderDetailViewModel.kt
// =============================================================================
// File:           mobile/android/feature/customerorder/src/main/kotlin/com/qervon/features/customerorder/OrderDetailViewModel.kt
// Project:        Qervon
// Author:         USDTG GROUP TECHNOLOGY LLC
// Developer:      Irfan Gedik
// Created Date:   2026-08-12
// Version:        0.1.0
//
// Description:
//   Polls `GET /v1/orders/{id}/tracking` and `GET /v1/customer/orders/{id}/eta`
//   every 5s while the order is `courier_assigned`/`in_transit`, and drives
//   cancel (`POST .../cancel`), rating (`POST .../rating`), and support
//   ticket creation.
//
// License:
//   Qervon License v1.0 — see LICENSE in the repository root.
// =============================================================================

package com.qervon.features.customerorder

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.qervon.core.common.QervonApiException
import com.qervon.core.common.model.EtaInfo
import com.qervon.core.common.model.LocationSnapshot
import com.qervon.core.common.model.Order
import com.qervon.core.common.model.OrderStatus
import com.qervon.core.network.QervonApi
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import javax.inject.Inject

private const val POLL_INTERVAL_MS = 5_000L

data class OrderDetailUiState(
    val order: Order? = null,
    val courierLocation: LocationSnapshot? = null,
    val eta: EtaInfo? = null,
    val isLoading: Boolean = false,
    val isSubmittingAction: Boolean = false,
    val errorMessage: String? = null,
    val infoMessage: String? = null,
) {
    val canCancel: Boolean get() = order?.status == OrderStatus.PENDING || order?.status == OrderStatus.COURIER_ASSIGNED
    val canRate: Boolean get() = order?.status == OrderStatus.DELIVERED
    val isTrackable: Boolean get() = order?.status == OrderStatus.COURIER_ASSIGNED || order?.status == OrderStatus.IN_TRANSIT
}

@HiltViewModel
class OrderDetailViewModel @Inject constructor(
    private val api: QervonApi,
    savedStateHandle: SavedStateHandle,
) : ViewModel() {

    private val orderId: String = savedStateHandle.get<String>("orderId").orEmpty()

    private val _uiState = MutableStateFlow(OrderDetailUiState())
    val uiState: StateFlow<OrderDetailUiState> = _uiState.asStateFlow()

    private var pollJob: Job? = null

    fun start() {
        refresh()
        if (pollJob?.isActive != true) {
            pollJob = viewModelScope.launch {
                while (isActive) {
                    delay(POLL_INTERVAL_MS)
                    refresh(showLoading = false)
                }
            }
        }
    }

    override fun onCleared() {
        pollJob?.cancel()
        super.onCleared()
    }

    private fun refresh(showLoading: Boolean = true) {
        viewModelScope.launch {
            if (showLoading) _uiState.value = _uiState.value.copy(isLoading = true)
            try {
                val orders = api.listCustomerOrders()
                val order = orders.firstOrNull { it.id == orderId }
                _uiState.value = _uiState.value.copy(order = order)
                if (order != null && (order.status == OrderStatus.COURIER_ASSIGNED || order.status == OrderStatus.IN_TRANSIT)) {
                    val eta = runCatching { api.getOrderEta(orderId) }.getOrNull()
                    val location = runCatching { api.getOrderTracking(orderId) }.getOrNull()
                    _uiState.value = _uiState.value.copy(eta = eta, courierLocation = location)
                }
            } catch (error: QervonApiException) {
                _uiState.value = _uiState.value.copy(errorMessage = error.message)
            } finally {
                _uiState.value = _uiState.value.copy(isLoading = false)
            }
        }
    }

    fun cancelOrder() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isSubmittingAction = true, errorMessage = null)
            try {
                val order = api.cancelOrder(orderId)
                _uiState.value = _uiState.value.copy(order = order, infoMessage = "Sipariş iptal edildi.")
            } catch (error: QervonApiException) {
                _uiState.value = _uiState.value.copy(errorMessage = error.message)
            } finally {
                _uiState.value = _uiState.value.copy(isSubmittingAction = false)
            }
        }
    }

    fun submitRating(stars: Int, comment: String?) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isSubmittingAction = true, errorMessage = null)
            try {
                api.rateOrder(orderId, stars, comment)
                _uiState.value = _uiState.value.copy(infoMessage = "Değerlendirmeniz için teşekkürler.")
            } catch (error: QervonApiException) {
                _uiState.value = _uiState.value.copy(errorMessage = error.message)
            } finally {
                _uiState.value = _uiState.value.copy(isSubmittingAction = false)
            }
        }
    }

    fun submitSupportTicket(subject: String, message: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isSubmittingAction = true, errorMessage = null)
            try {
                api.createSupportTicket(orderId, subject, message)
                _uiState.value = _uiState.value.copy(infoMessage = "Destek talebiniz iletildi.")
            } catch (error: QervonApiException) {
                _uiState.value = _uiState.value.copy(errorMessage = error.message)
            } finally {
                _uiState.value = _uiState.value.copy(isSubmittingAction = false)
            }
        }
    }
}
